router/oauthResponseHandler/oauthResponseHandler.go

```diff
@@ -0,0 +1,546 @@
+	ExpiredToken string `json:"expiredToken"`
+}
+
+func Init() {
+	configBEURL = backendconfig.GetConfigBackendURL()
+	pkgLogger = logger.NewLogger().Child("router").Child("OAuthResponseHandler")
+	loggerNm = "OAuthResponseHandler"
+	workspaceToken = backendconfig.GetWorkspaceToken()
+	destAuthInfoMap = make(map[string]*AuthResponse)
+	accountLockMap = make(map[string]*sync.RWMutex)
+	destLockMap = make(map[string]*sync.RWMutex)
+	lockMapMutex = &sync.RWMutex{}
+	refreshActiveMap = make(map[string]bool)
+	disableDestActiveMap = make(map[string]bool)
+}
+
+func (authErrHandler *OAuthErrResHandler) Setup() {
+	authErrHandler.logger = pkgLogger
+	authErrHandler.tr = &http.Transport{}
+	//This timeout is kind of modifiable & it seemed like 10 mins for this is too much!
+	authErrHandler.client = &http.Client{}
+	authErrHandler.destLockMap = destLockMap
+	authErrHandler.accountLockMap = accountLockMap
+	authErrHandler.lockMapWMutex = lockMapMutex
+	authErrHandler.destAuthInfoMap = destAuthInfoMap
+	authErrHandler.refreshActiveMap = refreshActiveMap
+	authErrHandler.disableDestActiveMap = disableDestActiveMap
+}
+
+func (authErrHandler *OAuthErrResHandler) RefreshToken(refTokenParams *RefreshTokenParams) (int, *AuthResponse) {
+	authStats := &OAuthStats{
+		id:              refTokenParams.AccountId,
+		workspaceId:     refTokenParams.WorkspaceId,
+		rudderCategory:  "destination",
+		statName:        "",
+		isCallToCpApi:   false,
+		authErrCategory: REFRESH_TOKEN,
+		errorMessage:    "",
+		destDefName:     refTokenParams.DestDefName,
+	}
+	return authErrHandler.GetTokenInfo(refTokenParams, "Refresh token", authStats)
+}
+
+func (authErrHandler *OAuthErrResHandler) FetchToken(fetchTokenParams *RefreshTokenParams) (int, *AuthResponse) {
+	authStats := &OAuthStats{
+		id:              fetchTokenParams.AccountId,
+		workspaceId:     fetchTokenParams.WorkspaceId,
+		rudderCategory:  "destination",
+		statName:        "",
+		isCallToCpApi:   false,
+		authErrCategory: "",
+		errorMessage:    "",
+		destDefName:     fetchTokenParams.DestDefName,
+		isTokenFetch:    true,
+	}
+	return authErrHandler.GetTokenInfo(fetchTokenParams, "Fetch token", authStats)
+}
+
+func (authErrHandler *OAuthErrResHandler) GetTokenInfo(refTokenParams *RefreshTokenParams, logTypeName string, authStats *OAuthStats) (int, *AuthResponse) {
+
+	startTime := time.Now()
+	defer func() {
+		authStats.statName = fmt.Sprintf("%v_total_req_latency", refTokenParams.EventNamePrefix)
+		authStats.isCallToCpApi = false
+		authStats.SendTimerStats(startTime)
+	}()
+	resMgrErr := authErrHandler.NewMutex(refTokenParams.AccountId, REFRESH_TOKEN)
+	if resMgrErr != nil {
+		panic(resMgrErr)
+	}
+
+	refTokenBody := RefreshTokenBodyParams{}
+	if router_utils.IsNotEmptyString(refTokenParams.AccessToken) {
+		refTokenBody = RefreshTokenBodyParams{
+			HasExpired:   true,
+			ExpiredToken: refTokenParams.AccessToken,
+		}
+	}
+	authErrHandler.accountLockMap[refTokenParams.AccountId].RLock()
```

Comment:
This map read is not thread-safe. You need to obtain a read lock:
```
	resHandler.lockMapWMutex.Lock()
```

Reply:
If I am understanding correctly, Is this what you're trying to say ?
```golang
authErrHandler.lockMapWMutex.RLock()
authErrHandler.accountLockMap[refTokenParams.AccountId].RLock()
// After all the reads are complete
authErrHandler.lockMapWMutex.RUnlock()
```